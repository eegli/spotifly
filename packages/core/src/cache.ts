/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/no-empty-function */

export enum CacheEntity {
  Artist = 'artist',
  Track = 'track',
}

export function Cacheable<T extends unknown[]>(
  entity: CacheEntity,
  debug = false
) {
  return function cacheMeBaby(
    _: unknown,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const log = debug ? console.log : () => {};
    const original = descriptor.value;
    descriptor.value = async function (...args: string[]) {
      if (!Array.isArray(args) || !args.every(arg => typeof arg === 'string')) {
        throw new Error(
          `Decorator applied on uncacheable method ${propertyKey}`
        );
      }
      const { cached, uncached } = cache.filterCached(entity, args);

      if (uncached.length === 0) {
        log('Reading all items from cache');
        return Promise.resolve(cached);
      }

      log(
        `Found ${cached.length}/${args.length} entries in cache, optimizing query...`
      );
      args = uncached;

      const result: T[] = await original.call(this, ...args);

      log(`Putting ${result.length} new item(s) into cache`);
      for (let i = 0; i < result.length; i++) {
        cache.put(entity, [uncached[i], result[i]]);
      }
      log(`Appending ${cached.length} item(s) to result from cache`);
      result.push(...cached);
      return Promise.resolve(result);
    };
  };
}

export type FilteredCache = {
  cached: any[];
  uncached: string[];
};

class Cache {
  #tracks = new Map<string, any>();
  #artists = new Map<string, any>();

  #enabled = true;

  #findBin(entity: CacheEntity) {
    switch (entity) {
      case CacheEntity.Track:
        return this.#tracks;
      case CacheEntity.Artist:
        return this.#artists;
    }
  }

  public disable() {
    this.#enabled = false;
  }

  public put(entity: CacheEntity, item: [string, any]): void {
    const bin = this.#findBin(entity);
    bin.set(item[0], item[1]);
  }

  public filterCached(entity: CacheEntity, ids: string[]): FilteredCache {
    if (!this.#enabled) return { cached: [], uncached: ids };
    const bin = this.#findBin(entity);
    return ids.reduce(
      (acc, curr) => {
        const item = bin.get(curr);
        if (item) {
          acc.cached.push(item);
        } else {
          acc.uncached.push(curr);
        }
        return acc;
      },
      {
        cached: [],
        uncached: [],
      } as FilteredCache
    );
  }
}

export const cache = new Cache();
