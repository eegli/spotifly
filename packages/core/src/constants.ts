export class Constants {
  public static readonly ApiBase = 'https://api.spotify.com/v1';
  public static readonly UserSavedTracks = 'me/tracks';
  public static readonly AudioFeatures = 'audio-features';
  public static readonly MultipleArtists = 'artists';
}
