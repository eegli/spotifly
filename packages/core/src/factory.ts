import { Params } from './models/params';
import type { AnyObject, AsyncFn, OmitFromAsyncFnParams } from './types';

type PagingObject<T> = SpotifyApi.PagingObject<T>;

type PaginationParams = Pick<Params, 'limit' | 'offset'>;

export function forPaginated<
  F extends AsyncFn<PagingObject<unknown>, string, PaginationParams>,
  R extends Awaited<ReturnType<F>>
>(getFn: F, limit: number) {
  return function (...args: OmitFromAsyncFnParams<F, PaginationParams>) {
    return async function (cb?: (param: R) => unknown): Promise<R[]> {
      const [arg1, arg2] = args;
      let nextPage: string | null = null;
      let offset = 0;
      const responses = [];
      do {
        const response = (await getFn(arg1, {
          ...arg2,
          limit,
          offset,
        })) as R;
        responses.push(response);
        nextPage = response.data.next;
        offset += limit;
        if (cb) await cb(response);
      } while (nextPage);
      return responses;
    };
  };
}

export function forLimited<
  F extends AsyncFn<unknown, string[], AnyObject>,
  R extends Awaited<ReturnType<F>>
>(getFn: F, limit: number) {
  return function (...args: Parameters<F>) {
    const [ids, rest] = args;
    return async function (cb?: (param: R) => unknown): Promise<R[]> {
      const responses: R[] = [];
      for (let i = 0; i < ids.length; i += limit) {
        const chunk = ids.slice(i, i + limit);
        const response = (await getFn(chunk, rest)) as R;
        if (cb) await cb(response);
        responses.push(response);
      }
      return responses;
    };
  };
}
