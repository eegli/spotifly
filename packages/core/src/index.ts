import { AuthProvider, AuthProviderOptions } from './provider';

import Albums from './models/albums';
import Artists from './models/artists';
import Categories from './models/categories';
import Episodes from './models/episodes';
import future from './models/future';
import Genres from './models/genres';
import Markets from './models/markets';
import Player from './models/player';
import Playlists from './models/playlists';
import Search from './models/search';
import Shows from './models/shows';
import Tracks from './models/tracks';
import Users from './models/users';

export function initialize(authOptions: AuthProviderOptions) {
  const provider = new AuthProvider(authOptions);
  return {
    Albums: Albums(provider),
    Artists: Artists(provider),
    Categories: Categories(provider),
    Episodes: Episodes(provider),
    Genres: Genres(provider),
    Markets: Markets(provider),
    Player: Player(provider),
    Playlists: Playlists(provider),
    Search: Search(provider),
    Shows: Shows(provider),
    Tracks: Tracks(provider),
    Users: Users(provider),
    future: future(provider),
  } as const;
}

export type SpotifyClient = ReturnType<typeof initialize>;

export { isError } from './request';
export type { DataPromise, DataResponse } from './types';
export type { AuthProviderOptions };
