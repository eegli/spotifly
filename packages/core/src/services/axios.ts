import axios from 'axios';

export const reqInterceptor: Parameters<typeof axios.interceptors.request.use> =
  [
    // Do something before request is sent
    config => config,
    // Do something with request error
    // Apparently, the error handler is used if the interceptor throws:
    // https://github.com/axios/axios/issues/2509
    error => Promise.reject(error),
  ];

export const resInterceptor: Parameters<
  typeof axios.interceptors.response.use
> = [
  response => response,
  async error => {
    if (error.response && error.response.status === 401) {
      // TODO
      console.log(error);
    }
    return Promise.reject(error);
  },
];
