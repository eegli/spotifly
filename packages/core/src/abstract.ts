import axios, { AxiosInstance } from 'axios';

export interface RequestAble {
  accessToken: string;
}

export class Requester {
  protected request: AxiosInstance;
  protected readonly BASE_URL_V1 = 'https://api.spotify.com/v1';

  constructor(token: string) {
    this.request = axios.create({
      baseURL: this.BASE_URL_V1,
      headers: { Authorization: `Bearer ${token}` },
    });
  }
}

/* export class Configurable<T> {
  private config: T;

  constructor(config: T) {
    this.config = config;
  }

  public update(config: Partial<T>) {
    this.config = { ...this.config, ...config };
  }

  public get get() {
    return this.config;
  }
}
 */
export abstract class SpotifyKind<C> extends Requester {
  protected options: RequestAble;
  public abstract get items(): C;

  // public abstract collect(...args: unknown[]): K;

  constructor(options: RequestAble) {
    super(options.accessToken);
    this.options = options;
  }
}
