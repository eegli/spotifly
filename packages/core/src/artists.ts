import { RequestAble, SpotifyKind } from './abstract';
import { MapCollection } from './types';
import { chunkify } from './utils';

export class Artists extends SpotifyKind<SpotifyApi.ArtistObjectFull[]> {
  private readonly MULTIPLE_ARTISTS = '/artists';
  private readonly CHUNK_SIZE = 50;

  private collection: MapCollection<SpotifyApi.ArtistObjectFull> = new Map();
  private artistIds: Set<string> = new Set();

  constructor(opts: RequestAble, id?: string, ...ids: string[]) {
    super(opts);
    if (id) {
      this.add(id, ...ids);
    }
  }

  add(id: string, ...ids: string[]) {
    [id, ...ids].forEach(id => this.artistIds.add(id));
  }

  get items() {
    return Array.from(this.collection.values());
  }

  async get(id: string, ...ids: string[]): Promise<void> {
    for await (const chunk of chunkify([id, ...ids], this.CHUNK_SIZE)) {
      const { data } =
        await this.request.get<SpotifyApi.MultipleArtistsResponse>(
          this.BASE_URL_V1 + this.MULTIPLE_ARTISTS,
          {
            params: {
              ids: chunk.join(','),
            },
          }
        );
      data.artists.forEach(artist => {
        this.collection.set(artist.id, artist);
      });
    }
  }
}
