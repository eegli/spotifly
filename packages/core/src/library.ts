import { RequestAble, SpotifyKind } from './abstract';
import { Artists } from './artists';
import { Constants } from './constants';
import { MapCollection, SavedTrack, UsersSavedTracks } from './types';

export class Library extends SpotifyKind<SavedTrack[]> {
  private collection: MapCollection<SavedTrack> = new Map();

  private readonly limit = 50;
  private pagination = {
    nextUrl: <string | null>Constants.UserSavedTracks,
    start: 0,
    end: this.limit,
  };

  private artists: Artists;

  constructor(opts: RequestAble) {
    super(opts);
    this.artists = new Artists(opts);
  }

  get items() {
    return Array.from(this.collection.values());
  }

  public async get(): Promise<this> {
    this.collection.clear();
    while (this.pagination.nextUrl) {
      const { data }: UsersSavedTracks = await this.request.get(
        this.pagination.nextUrl,
        {
          params: {
            limit: this.limit,
          },
        }
      );
      data.items.forEach(savedTrack => {
        this.collection.set(savedTrack.track.id, savedTrack);
        savedTrack.track.artists.forEach(({ id }) => {
          this.artists.add(id);
        });
      });
      this.pagination.nextUrl = data.next;
      this.pagination.start = this.pagination.end;
      this.pagination.end += data.limit;
    }

    return this;
  }

  public collect(type: 'light' | 'full') {
    if (type === 'full') {
      return Array.from(this.collection.values()).map(track => {
        return track;
      });
    }
    return [];
  }
}
