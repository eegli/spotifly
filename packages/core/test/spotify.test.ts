import axios from 'axios';
import { Library } from '../src/library';
import multipleArtistsResponse from './payloads/multiple_artists.json';
import userSavedTracksRespone from './payloads/user_saved_tracks.json';
jest.mock('axios');

const mockAxios = axios as jest.Mocked<typeof axios>;
mockAxios.create.mockReturnThis();

describe('Spotify', () => {
  test('config update', async () => {
    const lib = new Library({ token: '123' });
    lib.config.update({ token: 'new' });
    mockAxios.get.mockResolvedValueOnce({
      data: userSavedTracksRespone as SpotifyApi.SaveTracksForUserResponse,
    });
    await lib.get();
    expect(lib.tracks.length).toBe(userSavedTracksRespone.items.length);
    expect(lib.artists.).toBe(7);
    expect(lib.artists).toMatchSnapshot();
    mockAxios.get.mockResolvedValueOnce({
      data: multipleArtistsResponse as SpotifyApi.MultipleArtistsResponse,
    });
    await lib.enrichGenres();
    
  });
});
