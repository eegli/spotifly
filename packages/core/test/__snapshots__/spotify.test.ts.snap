// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spotify config update 1`] = `
Array [
  "7FtCyCJCJaxabYO7Uyda5B",
  "6PIxyj355iDhl3wQSsHXPE",
  "1BPsbKiNgav1TY1ITIU8C5",
  "3q4qbeWVDpN7LcwU6db2a7",
  "6QF9jZk7NKic4H0A01lGWw",
  "0k3dnuEr9LghUE4jxS0PT4",
  "5dScLouYq1Xn2eUlzlpusx",
]
`;
